applications/ContactStructuralMechanicsApplication/custom_utilities/interface_preprocess.cpp

```diff
@@ -29,220 +29,236 @@
+    const std::string& final_string = ThisParameters["final_string"].GetString();
+    const bool& simplest_geometry = ThisParameters["simplify_geometry"].GetBool();
+    
+    unsigned int cond_counter = 0;
+    
+    // We reorder the conditions
+    unsigned int cond_id = ReorderConditions();
 
-        // Check that we actually found something
-        if( NodesCounter == 0)
+    // Generate Conditions from original the faces that can be considered interface
+    for (ModelPart::ElementsContainerType::const_iterator it_elem = rOriginPart.ElementsBegin(); it_elem != rOriginPart.ElementsEnd(); ++it_elem)
+    {          
+        GeometryType& this_geometry = it_elem->GetGeometry();
+        
+        if (this_geometry.LocalSpaceDimension() == 3)
         {
-            KRATOS_ERROR << "No interface nodes found. Please check that nodes on both sides of the interface have been assigned Is(INTERFACE) = true." << std::endl;
+            for (unsigned int i_face = 0; i_face < this_geometry.FacesNumber(); ++i_face)
+            {
+                GenerateFaceCondition(rInterfacePart, *(it_elem.base()), this_geometry.Faces()[i_face], condition_name, final_string, simplest_geometry, cond_counter, cond_id);
+            }
         }
-        if( CondCounter == 0)
+        else
         {
-            KRATOS_ERROR << "No interface conditions found. Please check that nodes on both sides of the interface have been assigned Is(INTERFACE) = true and that the contact surfaces have been assigned conditions." << std::endl;
+            GenerateFaceCondition(rInterfacePart, *(it_elem.base()), this_geometry, condition_name, final_string, simplest_geometry, cond_counter, cond_id);
         }
     }
     
-    /***********************************************************************************/
-    /***********************************************************************************/
+    // NOTE: Reorder ID if parallellization
+    
+    const unsigned int num_nodes = static_cast<int>(mrMainModelPart.Nodes().size());
+    PrintNodesAndConditions(num_nodes, cond_counter);
     
-    unsigned int InterfacePreprocessCondition::ReorderConditions()
+    KRATOS_CATCH("");
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void InterfacePreprocessCondition::CreateNewCondition(
+        Element::Pointer rpElem,
+        Geometry<Node<3> > & rGeometry,
+        const unsigned int CondId,
+        const std::string& ConditionName
+        )
+{
+    KRATOS_TRY;
+
+    Condition const& rCondition = KratosComponents<Condition>::Get(ConditionName); 
+    Condition::Pointer p_cond = Condition::Pointer(rCondition.Create(CondId, rGeometry, rpElem->pGetProperties()));
+    GeometryType& this_geometry = p_cond->GetGeometry();
+    mrMainModelPart.AddCondition(p_cond);
+    if (ConditionName.find("Mortar") != std::string::npos)
```

Comment:
it is NOT a good idea to operate with strings within a function that is used many times. It is EXTREMELY slow

Reply:
Ok